src/Angor/Client/Components/MessageComponent.razor

```diff
@@ -0,0 +1,402 @@
+    private bool _needsScroll = false;
+    private int _messageCount = 0;
+
+    protected override async Task OnInitializedAsync()
+    {
+        await base.OnInitializedAsync();
+
+        if (string.IsNullOrEmpty(ContactHexPub) && !string.IsNullOrEmpty(ContactNpub))
+        {
+            ContactHexPub = NostrHelper.ConvertBech32ToHex(ContactNpub);
+        }
+
+        if (!string.IsNullOrEmpty(ContactHexPub) && string.IsNullOrEmpty(ContactNpub))
+        {
+            try { ContactNpub = NostrHelper.ConvertHexToNpub(ContactHexPub) ?? ContactHexPub; }
+            catch { ContactNpub = ContactHexPub; }
+        }
+
+        MessageService.OnChange += HandleMessageServiceChange;
+    }
+
+    protected override async Task OnParametersSetAsync()
+    {
+        await base.OnParametersSetAsync();
+
+        if (!string.IsNullOrEmpty(CurrentUserPrivateKeyHex) && !string.IsNullOrEmpty(CurrentUserNpub) && !string.IsNullOrEmpty(ContactHexPub))
+        {
+            await InitializeMessaging();
+        }
+    }
+
+    protected override async Task OnAfterRenderAsync(bool firstRender)
+    {
+        if (_needsScroll)
+        {
+            await JS.InvokeVoidAsync("scrollToBottom", messagesContainerRef);
+            _needsScroll = false;
+        }
+    }
+
+    private async Task InitializeMessaging()
+    {
+        if (string.IsNullOrEmpty(CurrentUserPrivateKeyHex) ||
+            string.IsNullOrEmpty(CurrentUserNpub) ||
+            string.IsNullOrEmpty(ContactHexPub))
+        {
+            await OnNotification.InvokeAsync("Missing required message information (keys/contact).");
+            return;
+        }
+
+        try
+        {
+
+            await MessageService.InitializeAsync(CurrentUserPrivateKeyHex, CurrentUserNpub, ContactHexPub);
+            _needsScroll = true;
+        }
+        catch (Exception ex)
+        {
+            Logger.LogError(ex, "Failed to initialize messaging service.");
+            await OnNotification.InvokeAsync($"Error initializing messages: {ex.Message}");
+        }
+
+    }
+
+    private async Task SendMessageAsync()
+    {
+        if (string.IsNullOrWhiteSpace(newMessage) || MessageService.IsSendingMessage) return;
+
+        string messageToSend = newMessage;
+        newMessage = "";
+
+        try
+        {
+            await MessageService.SendMessageAsync(messageToSend);
+            _needsScroll = true;
+        }
+        catch (Exception ex)
+        {
+            Logger.LogError(ex, "Failed to send message.");
+            await OnNotification.InvokeAsync($"Failed to send message: {ex.Message}");
+            newMessage = messageToSend;
+        }
+    }
+
+    private async Task RefreshMessagesAsync()
+    {
+        if (MessageService.IsRefreshing) return;
+
+        showActionsMenu = false;
+
+        try
+        {
+            await MessageService.RefreshMessagesAsync();
+            _needsScroll = true;
+            await OnNotification.InvokeAsync("Messages refreshed.");
```

Comment:
This we don't need for each refresh, 
better is a small spinner on the popup itself.